core/src/main/scala/org/typelevel/otel4s/trace/Span.scala

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright 2022 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s
+package trace
+
+import scala.concurrent.duration.FiniteDuration
+
+/** The API to trace an operation.
+  *
+  * There are three types of span: [[Span.Manual]], [[Span.Auto]], [[Span.Res]].
+  *
+  * ==[[Span.Manual]]==
+  * The manual span requires to be ended '''explicitly'''. Manual span can be
+  * used when it's necessary to end a span outside of the resource scope (i.e.
+  * async callback). Make sure the span is ended properly.
+  *
+  * Leaked span:
+  * {{{
+  * val tracer: Tracer[F] = ???
+  * val leaked: F[Unit] =
+  *   tracer.spanBuilder("manual-span").createManual.use { span =>
+  *     span.setStatus(Status.Ok, "all good")
+  *   }
+  * }}}
+  *
+  * Properly ended span:
+  * {{{
+  * val tracer: Tracer[F] = ???
+  * val ok: F[Unit] =
+  *   tracer.spanBuilder("manual-span").createManual.use { span =>
+  *     span.setStatus(Status.Ok, "all good") >> span.end
+  *   }
+  * }}}
+  *
+  * ==[[Span.Auto]]==
+  * Unlike [[Span.Manual]] the auto span has a fully managed lifecycle. That
+  * means the span is started upon resource allocation and ended upon
+  * finalization.
+  *
+  * Automatically ended span:
+  * {{{
+  * val tracer: Tracer[F] = ???
+  * val ok: F[Unit] =
+  *   tracer.spanBuilder("manual-span").createAuto.use { span =>
+  *     span.setStatus(Status.Ok, "all good")
+  *   }
+  * }}}
+  *
+  * ==[[Span.Res]]==
+  * The behaviour and lifecycle management are identical to [[Span.Auto]]. The
+  * allocation and release stages of a supplied resource are traced by separate
+  * spans. Carries a value of a wrapped resource.
+  *
+  * The structure of the inner spans:
+  * {{{
+  * > span-name
+  *   > acquire
+  *   > use
+  *   > release
+  * }}}
+  *
+  * Resource span:
+  * {{{
+  * val tracer: Tracer[F] = ???
+  * val resource: Resource[F, String] = Resource.eval(Sync[F].delay("string"))
+  * val ok: F[Unit] =
+  *   tracer.spanBuilder("manual-span").createRes(resource).use { case span @ Span.Res(value) =>
+  *     span.setStatus(Status.Ok, s"all good. resource value: $${value}")
+  *   }
+  * }}}
+  */
+trait Span[F[_]] {
+
+  /** Returns the [[SpanContext]] associated with this span.
+    *
+    * Returns `None` if the span is invalid or no-op.
+    */
+  def context: Option[SpanContext]
```

Comment:
In t4c we have an `invalid` instance of `SpanContext`. I think I like using an `Option` here better, but we might find that the Java implementation needs an `invalid` instance

Reply:
Yeah, I'm scratching my head about this topic. 

Official spec https://opentelemetry.io/docs/reference/specification/compatibility/opentracing/#get-the-active-span.


According to the OpenTelemetry specification, the invalid context can be described as
```scala
SpanContext { 
  traceId = "00000000000000000000000000000000", 
  spanId = "0000000000000000"
} 
```

If I get it right, `invalid` and `noop` spans are identical from the OpenTelemetry perspective. They should not produce child spans at all.

I see three states of the `SpanContext`:
1) Valid
2) Invalid (i.e. cannot create a span from remote `traceId` and `spanId`)
3) Noop

Perhaps it's better to stick with OpenTelemetry design and always return `SpanContext` because, well, the context is always there:
```scala
def context: SpanContext
```

Then a library user should verify the validity of the context if necessary.

Reply:
While I like the `Option`, you're right it's probably a good idea to stay close to the spec.

> Then a library user should verify the validity of the context if necessary.

This hasn't hurt us in t4c as far as I'm aware. No one has needed to differentiate between a no-op and invalid context.